synthetics-stats/src/entities/volume.ts

```diff
@@ -0,0 +1,119 @@
+  dailyVolumeInfo.volumeUsd = dailyVolumeInfo.volumeUsd.plus(volume);
+  totalVolumeInfo.volumeUsd = totalVolumeInfo.volumeUsd.plus(volume);
+
+  if (type == "swap") {
+    hourlyVolumeInfo.swapVolumeUsd = hourlyVolumeInfo.swapVolumeUsd.plus(volume);
+    dailyVolumeInfo.swapVolumeUsd = dailyVolumeInfo.swapVolumeUsd.plus(volume);
+    totalVolumeInfo.swapVolumeUsd = totalVolumeInfo.swapVolumeUsd.plus(volume);
+  } else if (type == "margin") {
+    hourlyVolumeInfo.marginVolumeUsd = hourlyVolumeInfo.marginVolumeUsd.plus(volume);
+    dailyVolumeInfo.marginVolumeUsd = dailyVolumeInfo.marginVolumeUsd.plus(volume);
+    totalVolumeInfo.marginVolumeUsd = totalVolumeInfo.marginVolumeUsd.plus(volume);
+  }
+
+  hourlyVolumeInfo.save();
+  dailyVolumeInfo.save();
+  totalVolumeInfo.save();
+}
+
+function getOrCreateVolumeInfo(timestamp: i32, period: string): VolumeInfo {
+  let timestampGroup = timestampToPeriodStart(timestamp, period);
+  let volumeId = period === "total" ? "total" : timestampGroup.toString();
+  let volumeInfo = VolumeInfo.load(volumeId);
+
+  if (volumeInfo === null) {
+    volumeInfo = new VolumeInfo(volumeId);
+    volumeInfo.period = period;
+    volumeInfo.volumeUsd = BigInt.fromI32(0);
+    volumeInfo.swapVolumeUsd = BigInt.fromI32(0);
+    volumeInfo.marginVolumeUsd = BigInt.fromI32(0);
+  }
+  return volumeInfo as VolumeInfo;
+}
+
+export function saveSwapVolumeInfo(eventData: EventData, transaction: Transaction): void {
+  let PERIODS = new Array<string>(3);
+  PERIODS[0] = "1h";
+  PERIODS[1] = "1d";
+  PERIODS[2] = "total";
+
+  let timestamp = transaction.timestamp;
+  let tokenIn = eventData.getAddressItemString("tokenIn")!;
+  let tokenOut = eventData.getAddressItemString("tokenOut")!;
+  let amountIn = eventData.getUintItem("amountIn")!;
+  let tokenInPrice = eventData.getUintItem("tokenInPrice")!;
+  let volumeUsd = amountIn.times(tokenInPrice);
+
+  createOrUpdateSwapVolumeInfo(PERIODS, timestamp, volumeUsd, tokenIn, tokenOut);
+}
+
+function createOrUpdateSwapVolumeInfo(periodsToCreate: string[], timestamp: i32, volumeUsd: BigInt, tokenIn: string, tokenOut: string): void {
+  for (let i = 0; i < periodsToCreate.length; i++) {
+    let period = periodsToCreate[i];
+    let timestampGroup = timestampToPeriodStart(timestamp, period);
+    let id = period === "total" ? "total" : getVolumeInfoByTokenId(tokenIn, tokenOut) + ":" + timestampGroup.toString();
+    let volumeInfo = SwapVolumeInfo.load(id);
+
+    if (volumeInfo === null) {
+      volumeInfo = new SwapVolumeInfo(id);
+      volumeInfo.tokenIn = tokenIn;
+      volumeInfo.tokenOut = tokenOut;
+      volumeInfo.timestamp =timestampGroup;
+      volumeInfo.period = period;
+      volumeInfo.volumeUsd = BigInt.fromI32(0);
+    }
+
+    volumeInfo.volumeUsd = volumeInfo.volumeUsd.plus(volumeUsd);
+    volumeInfo.save();
+  }
+}
+
+export function savePositionVolumeInfo(eventData: EventData, transaction: Transaction): void {
```

Comment:
i though to use already created entities like PositionIncrease / PositionDecrease because we don't have any types for eventData and this is a bit unclear for which event this function is designed

but it's not critial and up to you, your approach is okay too in this case

Reply:
I have refactored the function to be a bit more simpler.